doc/bmc-info.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.33.
.TH BMC-INFO "1" "February 2004" "bmc-info 0.1.0" "User Commands"
.SH NAME
bmc-info \- manual page for bmc-info.  BMC Info 0.1.0
.SH DESCRIPTION
Displays BMC Information.
.SH OPTIONS
.TP
\fB\-u\fR, \fB\-\-usage\fR
Usage message
.TP
\fB\-h\fR, \fB\-\-help\fR
Show help
.TP
\fB\-V\fR, \fB\-\-version\fR
Show version
.SH "REPORTING BUGS"
Report bugs to <ab@gnu.org.in>.
.SH COPYRIGHT
Copyright \(co 2003-2004 FreeIPMI Core Team.  
This program is free software; you may redistribute it under the terms of
the GNU General Public License.  This program has absolutely no warranty.
.SH "SEE ALSO"
The full documentation for
.B bmc-info
is maintained as a Texinfo manual.  If the
.B info
and
.B bmc-info
programs are properly installed at your site, the command
.IP
.B info bmc-info
.PP
should give you access to the complete manual.


NEWS
FreeIPMI 1.7.1 - XX/XX/XX
-------------------------

In 2020 the IPMI standard was frozen.  So a decision was made to try
and make the FreeIPMI 1.7.X line the last stable line in FreeIPMI.  A
number of decisions were made to remove legacy support and fix minor issues,
even at the cost of breaking backwards compatibility.

Updated Features:
-----------------

Globally
--------
o Internally allow hostnames to be up to 256 characters (previously only 64).

Libraries
---------
o libfreeipmi/sensor-read: Add error for privilege level insufficient.
o In libfreeipmi, support Intel S2600BPB OEM SEL interpretations.
o Significant refactoring of Intel OEM extensions.  Names of
  variables, macros, etc. may have changed.

Tools
-----
o In ipmi-sel, support Intel S2600BPB OEM SEL interpretations.
o In ipmi-config, support configuration of Maximum_Privilege_Cipher_Suite_Id_16.
o In ipmi-sensors, support new "altbridging" workaround for Lenovo motherboards.

Smaller Bug Fixes
-----------------

o In ipmi-config, drop support for
  Rmcpplus_Conf_Privilege:Maximum_Privilege_Cipher_Suite_Id_16.  It is
  not configurable.

o Globally, fix potential output sorting error with
  --consolidate-output.  Could alter output differently than before.

o Fix many minor output errors (i.e. typos, space alignment, lack of
  newline, etc.).  Could cause issues for those who script against
  output.

o In libfreeipmi, add missing fields for BMC Message Channel SDR
  record format.

Remove Large Amount of Legacy Support
-------------------------------------

Remove backwards compatibility, legacy output/options/fields/files, or
deprecated options.  All verified to have been gone for atleast 10
years.

For details of changes see NEWS.errata

Typo Fixes:
-----------

Fix large number of typo and minor output formattings.  They have the
potential to break against existing builds or scripts or configs that
relied on previous expected output.

See NEWS.errata for details

FreeIPMI 1.6.1 - 02/02/18
-------------------------
o Add IPv6 hostname support to FreeIPMI, all of FreeIPMI can now take
  IPv6 addresses as inputs to "host" parameters, options, or inputs.
o Support significant portions of IPMI IPv6 configuration in
  libfreeipmi.
o Add --no-session option in ipmi-raw.
o Add SDR cache options to ipmi-config.
o Legacy -f short option for --flush-cache and -Q short option for
  quiet-cache.  Backwards compatible for tools that supported it
  before.
o In ipmi-oem, support Gigabyte get-bmc-services and set-bmc-services.
o Various performance improvements:
  - Remove excessive calls to secure_memset to clear memory.
  - Remove excessive memsets and clears of data.
  - Remove unnecessary "double input checks".
  - Remove expensive input checks in libfreeipmi fiid library.
    Fallout from this may include FIID_ERR_FIELD_NOT_FOUND errors in
    different fiid functions.
  - Remove unnecessary input checks in libfreeipmi fiid library.
  - Add recent 'lookups' of fields in fiid library to internal cache.
o Various minor fixes/improvements
  - Update libfreeipmi core API to use poll() instead of select(), to
    avoid issues with applications with a high number of threads.

FreeIPMI 1.5.7 - 08/16/17
-------------------------
o In libipmimonitoring, fix several mem-leak corner cases.

FreeIPMI 1.5.6 - 07/05/17
-------------------------
o In libfreeipmi locate (used by ipmi-locate), use DMI firmware in
  sysfs if it exists.
o Minor mem-leak corner case fix in libfreeipmi.
o Minor documentation fixes.
o Minor error message clarifications.

FreeIPMI 1.5.5 - 11/02/16
-------------------------
o Fix invalid flag clear in libipmiconsole that can lead to a potential
  double close on a file descriptor.
o Support Supermicro H8SGL-F OEM sensors and events.

FreeIPMI 1.5.4 - 09/30/16
-------------------------
o Various changes/fixes in libipmiconsole
  - If user retrieves file descriptor from ipmiconsole_ctx_fd, user is
    required to close it.  ipmiconsole_ctx_destroy no longer closes
    it.  This is to avoid a potential double close which can be a
    problem for multithreaded applications.  This is a change in
    behavior, but we do not believe this will affect most applications
    since most users close the file descriptor under most scenarios
    anyways.
  - ipmiconsole_ctx_destroy() should now be called to free resources
    even if ipmiconsole_engine_teardown() has been called.  This has
    been done to create consistent behavior in the API and avoid a
    former segfault possibility.  This is a change in behavior, but we
    do not believe this will affect most applications since
    ipmiconsole_engine_teardown() is only called when an application
    is being shutdown.
  - The use of IPMICONSOLE_ENGINE_CLOSE_FD has been clarified in the
    header file.  Some of the prior text was unclear.  Behavior has
    not been changed.
o In ipmi-oem, support Intel get-bmc-services and set-bmc-services commands.
o In ipmi-oem, support Gigabyte get-nic-mode and set-nic-mode commands.
o Support Gigabyte MD90-FS0-ZB OEM SEL events.

FreeIPMI 1.5.3 - 08/01/16
-------------------------
o Update Fujitsu OEM extension product id support range.
o In ipmiconsole, support --debugfile option for more difficult
  debugging scenarios.
o In all sensor related tools, increase max number of record ids and
  sensor types that can be handled.
o Fix several invalid completion code errors/values.

FreeIPMI 1.5.2 - 05/24/16
-------------------------
o Update configuration file parsing code to handle values up to 1024
  characters in length.
o Set FD_CLOEXEC for ipmi driver device files opened within
  libfreeipmi.
o Support --read-fru, --write-fru, and --device-id option in
  bmc-device.
o Support --fru-file option in ipmi-fru.
o Various library updates in libfreeipmi to handle new features.

FreeIPMI 1.5.1 - 12/01/15
-------------------------
o Update Intel Node Manager support for 3.0 specification.
  - Update API and fiid templates in libfreeipmi.

fish/extensions/bmc-info.scm
;;; bmc-info.scm: Displays BMC Information
;;; authors: M.P.Anand Babu <ab@gnu.org.in> 
;;; Balamurugan <bala.a@californiadigital.com>

;;; This program is free software; you can redistribute it and/or
;;; modify it under the terms of the GNU General Public License as
;;; published by the Free Software Foundation; either version 2, or (at
;;; your option) any later version.
;;; 
;;; This program is distributed in the hope that it will be useful, but
;;; WITHOUT ANY WARRANTY; without even the implied warranty of
;;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
;;; General Public License for more details.
;;; 
;;; You should have received a copy of the GNU General Public License
;;; along with this program; if not, write to the Free Software
;;; Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA
;;; 
;;; bmc-info.scm should be automatically loaded thru init.scm

(use-modules (ice-9 getopt-long))

(define bmc-info-exit-status 0)

(define (bmc-info-display-usage)
  (begin 
    (display "Usage: bmc-info [-?V] [-D IPMIDRIVER] [-h IPMIHOST] [-u USERNAME]\n")
    (display "                [-p PASSWORD] [-a AUTHTYPE] [-l PRIVILEGE-LEVEL]\n")
    (display "                [--no-probing] [--driver-type=IPMIDRIVER]\n")
    (display "                [--driver-address=DRIVERADDR] [--driver-device=DEVICE]\n")
    (display "                [--packet-retry-timeout=TIMEOUT] [--packet-retry-max=COUNT]\n")
    (display "                [--hostname=IPMIHOST] [--username=USERNAME]\n")
    (display "                [--password=PASSWORD] [--auth-type=AUTHTYPE]\n")
    (display "                [--priv-level=PRIVILEGE-LEVEL] [--help] [--usage]\n")
    (display "                [--version]\n")))

(define (bmc-info-display-help)
  (begin 
    (display "Usage: bmc-info [OPTION...] \n")
    (display "bmc-info displays information about BMC.\n")
    (display "\n")
    (display "      --no-probing           Do not probe IPMI devices.\n")
    (display "  -D, --driver-type=IPMIDRIVER   Use this IPMIDRIVER instead of auto selection.\n")
    (display "                              Allowed values are KCS, SMIC, SSIF and LAN.\n")
    (display "      --driver-address=DRIVERADDR\n")
    (display "                             Use this DRIVERADDR address instead of probed\n")
    (display "                             one.\n")
    (display "      --driver-device=DEVICE Use this DEVICE for IPMI driver\n")
    (display "  --packet-retry-timeout=TIMEOUT\n")
    (display "                         Use TIMEOUT when reading LAN packets in UDM.\n")
    (display "  --packet-retry-max=COUNT   Use COUNT retries when reading LAN packets get\n")
    (display "                         timed out in UDM.\n")
    (display "  -h, --host=IPMIHOST        Connect to IPMIHOST.\n")
    (display "  -u, --username=USERNAME    Use USERNAME instead of NULL.  Maximum USERNAME\n")
    (display "                             length is 16.\n")
    (display "  -p, --password=PASSWORD    Use PASSWORD instead of NULL.  Maximum PASSWORD\n")
    (display "                             length is 16.\n")
    (display "  -a, --auth-type=AUTHTYPE   Use AUTHTYPE instead of NONE.  Allowed values are\n")
    (display "                             NONE, MD2, MD5, PLAIN and OEM.\n")
    (display "  -l, --priv-level=PRIVILEGE-LEVEL\n")
    (display "                             Use this PRIVILEGE-LEVEL instead of USER.  Allowed\n")
    (display "                             values are CALLBACK, USER, OPERATOR, ADMIN and\n")
    (display "                             OEM.\n")
    (display "  -?, --help                 Give this help list.\n")
    (display "      --usage                Give a short usage message.\n")
    (display "  -V, --version              Print program version.\n")
    (display "\n")
    (display "Mandatory or optional arguments to long options are also mandatory or optional\n")
    (display "for any corresponding short options.\n")
    (display "\n")
    (display "Report bugs to <freeipmi-devel@gnu.org>.\n")))

(define (bmc-info-display-version)
  (begin 
    (display (string-append "BMC Information [bmc-info-" (fi-version) "]\n"))
    (display "Copyright (C) 2003-2005 FreeIPMI Core Team\n")
    (display "This program is free software; you may redistribute it under the terms of\n")
    (display "the GNU General Public License.  This program has absolutely no warranty.\n")))

(define (bmc-info-argp args)
  (catch 'misc-error 
	 (lambda () 
	   (let* ((bmc-info-cmd-args '())
		  (option-spec '((no-probing     (single-char #\202) (value #f))
				 (driver-type    (single-char #\D)   (value #t))
				 (driver-address (single-char #\203) (value #t))
				 (driver-device  (single-char #\204) (value #t))
				 (packet-retry-timeout  (single-char #\205) (value #t))
				 (packet-retry-max      (single-char #\206) (value #t))
				 (host           (single-char #\h)   (value #t))
				 (username       (single-char #\u)   (value #t))
				 (password       (single-char #\p)   (value #t))
				 (auth-type      (single-char #\a)   (value #t))
				 (priv-level     (single-char #\l)   (value #t))
				 (help           (single-char #\?)   (value #f))
				 (usage          (single-char #\377) (value #f))
				 (version        (single-char #\V)   (value #f))))
		  (options (getopt-long args option-spec))
		  (no-probing     (option-ref options 'no-probing     #f))
		  (driver-type    (option-ref options 'driver-type    #f))
		  (driver-address (option-ref options 'driver-address #f))
		  (driver-device  (option-ref options 'driver-device  #f))
		  (retry-timeout  (option-ref options 'packet-retry-timeout #f))
		  (retry-max      (option-ref options 'packet-retry-max     #f))
		  (host           (option-ref options 'host           #f))
		  (username       (option-ref options 'username       #f))
		  (password       (option-ref options 'password       #f))
		  (auth-type      (option-ref options 'auth-type      #f))
		  (priv-level     (option-ref options 'priv-level     #f))
		  (help-wanted    (option-ref options 'help           #f))
		  (usage-wanted   (option-ref options 'usage          #f))
		  (version-wanted (option-ref options 'version        #f))
		  (extra-args     (option-ref options '()             #f)))
	     ;; extra arguments
	     (if (and (not (null? extra-args)) (list? bmc-info-cmd-args))
		 (begin 
		   (display "Usage: bmc-info [OPTION...] \n"
			    (current-error-port))
		   (display "Try `bmc-info --help' or `bmc-info --usage' for more information.\n"
			    (current-error-port))
		   (set! bmc-info-exit-status 64)
		   (set! bmc-info-cmd-args #f)))
	     ;; --no-probing (0)
	     (if (list? bmc-info-cmd-args)
		 (set! bmc-info-cmd-args (append bmc-info-cmd-args 
						 (list no-probing))))
	     ;; --driver-type (1)
	     (if (and (string? driver-type) (list? bmc-info-cmd-args))
		 (cond 
		  ((string-ci=? driver-type "lan")
		   (set! driver-type 1))
		  ((string-ci=? driver-type "kcs")
		   (set! driver-type 2))
		  ((string-ci=? driver-type "smic")
		   (set! driver-type 3))
		  ((string-ci=? driver-type "bt")
		   (set! driver-type 4))
		  ((string-ci=? driver-type "ssif")
		   (set! driver-type 5))
		  (else 
		   (begin 
		     (display "Usage: bmc-info [OPTION...] \n"
			      (current-error-port))
		     (display "Try `bmc-info --help' or `bmc-info --usage' for more information.\n"
			      (current-error-port))
		     (set! bmc-info-exit-status 64)
		     (set! bmc-info-cmd-args #f))))
		 (set! driver-type 0))
	     (if (list? bmc-info-cmd-args)
		 (set! bmc-info-cmd-args (append bmc-info-cmd-args 
						 (list driver-type))))
	     ;; driver-address (2)
	     (if (and (string? driver-address) (list? bmc-info-cmd-args))
		 (begin 
		   (set! driver-address (string->number driver-address))
		   (if (boolean? driver-address)
		       (begin 
			 (display "Usage: bmc-info [OPTION...] \n"
				  (current-error-port))
			 (display "Try `bmc-info --help' or `bmc-info --usage' for more information.\n"
				  (current-error-port))
			 (set! bmc-info-exit-status 64)
			 (set! bmc-info-cmd-args #f)))))
	     (if (list? bmc-info-cmd-args)
		 (set! bmc-info-cmd-args (append bmc-info-cmd-args 
						 (list driver-address))))
	     ;; --driver-device (3)
	     (if (list? bmc-info-cmd-args)
		 (set! bmc-info-cmd-args (append bmc-info-cmd-args 
						 (list driver-device))))
	     ;; --packet-retry-timeout (4)
	     (if (and (string? retry-timeout) (list? bmc-info-cmd-args))
		 (begin 
		   (set! retry-timeout (string->number retry-timeout))
		   (if (boolean? retry-timeout)
		       (begin 
			 (display "Usage: ipmi-sensors [OPTION...] \n"
				  (current-error-port))
			 (display "Try `ipmi-sensors --help' or `ipmi-sensors --usage' for more information.\n"
				  (current-error-port))
			 (set! sensor-exit-status 64)
			 (set! bmc-info-cmd-args #f)))))
	     (if (list? bmc-info-cmd-args)
		 (set! bmc-info-cmd-args (append bmc-info-cmd-args 
						 (list retry-timeout))))
	     ;; --packet-retry-max (5)
	     (if (and (string? retry-max) (list? bmc-info-cmd-args))
		 (begin 
		   (set! retry-max (string->number retry-max))
		   (if (boolean? retry-max)
		       (begin 
			 (display "Usage: ipmi-sensors [OPTION...] \n"
				  (current-error-port))
			 (display "Try `ipmi-sensors --help' or `ipmi-sensors --usage' for more information.\n"
				  (current-error-port))
			 (set! sensor-exit-status 64)
			 (set! bmc-info-cmd-args #f)))))
	     (if (list? bmc-info-cmd-args)
		 (set! bmc-info-cmd-args (append bmc-info-cmd-args 
						 (list retry-max))))
	     ;; --host (6)
	     (if (list? bmc-info-cmd-args)
		 (set! bmc-info-cmd-args (append bmc-info-cmd-args 
						 (list host))))
	     ;; --username (7)
	     (if (and (string? username) (list? bmc-info-cmd-args))
		 (begin 
		   (if (> (string-length username) 16)
		       (begin 
			 (display "Usage: bmc-info [OPTION...] \n"
				  (current-error-port))
			 (display "Try `bmc-info --help' or `bmc-info --usage' for more information.\n"
				  (current-error-port))
			 (set! bmc-info-exit-status 64)
			 (set! bmc-info-cmd-args #f)))))
	     (if (list? bmc-info-cmd-args)
		 (set! bmc-info-cmd-args (append bmc-info-cmd-args 
						 (list username))))
	     ;; --password (8)
	     (if (and (string? password) (list? bmc-info-cmd-args))
		 (begin 
		   (if (> (string-length password) 16)
		       (begin 
			 (display "Usage: bmc-info [OPTION...] \n"
				  (current-error-port))
			 (display "Try `bmc-info --help' or `bmc-info --usage' for more information.\n"
				  (current-error-port))
			 (set! bmc-info-exit-status 64)
			 (set! bmc-info-cmd-args #f)))))
	     (if (list? bmc-info-cmd-args)
		 (set! bmc-info-cmd-args (append bmc-info-cmd-args 
						 (list password))))
	     ;; --auth-type (9)
	     (if (and (string? auth-type) (list? bmc-info-cmd-args))
		 (cond 
		  ((string-ci=? auth-type "none")
		   (set! auth-type 0))
		  ((string-ci=? auth-type "md2")
		   (set! auth-type 1))
		  ((string-ci=? auth-type "md5")
		   (set! auth-type 2))
		  ((string-ci=? auth-type "plain")
		   (set! auth-type 4))
		  ((string-ci=? auth-type "oem")
		   (set! auth-type 5))
		  (else 
		   (begin 
		     (display "Usage: bmc-info [OPTION...] \n"
			      (current-error-port))
		     (display "Try `bmc-info --help' or `bmc-info --usage' for more information.\n"
			      (current-error-port))
		     (set! bmc-info-exit-status 64)
		     (set! bmc-info-cmd-args #f))))
		 (set! auth-type 0))
	     (if (list? bmc-info-cmd-args)
		 (set! bmc-info-cmd-args (append bmc-info-cmd-args 
						 (list auth-type))))
	     ;; --priv-level (10)
	     (if (and (string? priv-level) (list? bmc-info-cmd-args))
		 (cond 
		  ((string-ci=? priv-level "callback")
		   (set! priv-level 1))
		  ((string-ci=? priv-level "user")
		   (set! priv-level 2))
		  ((string-ci=? priv-level "operator")
		   (set! priv-level 3))
		  ((string-ci=? priv-level "admin")
		   (set! priv-level 4))
		  ((string-ci=? priv-level "oem")
		   (set! priv-level 5))
		  (else 
		   (begin 
		     (display "Usage: bmc-info [OPTION...] \n"
			      (current-error-port))
		     (display "Try `bmc-info --help' or `bmc-info --usage' for more information.\n"
			      (current-error-port))
		     (set! bmc-info-exit-status 64)
		     (set! bmc-info-cmd-args #f))))
		 (set! priv-level 2))
	     (if (list? bmc-info-cmd-args)
		 (set! bmc-info-cmd-args (append bmc-info-cmd-args 
						 (list priv-level))))
	     ;; --help (11)
	     (if (list? bmc-info-cmd-args)
		 (set! bmc-info-cmd-args (append bmc-info-cmd-args 
						 (list help-wanted))))
	     ;; --usage (12)
	     (if (list? bmc-info-cmd-args)
		 (set! bmc-info-cmd-args (append bmc-info-cmd-args 
						 (list usage-wanted))))
	     ;; --version (13)
	     (if (list? bmc-info-cmd-args)
		 (set! bmc-info-cmd-args (append bmc-info-cmd-args 
						 (list version-wanted))))
	     bmc-info-cmd-args))
	 (lambda (k args . opts)
	   (display "bmc-info: error: " (current-error-port))
	   (display (cadr opts) (current-error-port))
	   (display "\n" (current-error-port))
	   (display "Usage: bmc-info [OPTION...] \n"
		    (current-error-port))
	   (display "Try `bmc-info --help' or `bmc-info --usage' for more information.\n"
		    (current-error-port))
	   (set! bmc-info-exit-status 64)
	   #f)))

(define (bmc-info-get-help-option cmd-args)
  (list-ref cmd-args 11))

(define (bmc-info-get-usage-option cmd-args)
  (list-ref cmd-args 12))

(define (bmc-info-get-version-option cmd-args)
  (list-ref cmd-args 13))

(define (bmc-info-main cmd-args)
  (cond 
   ((bmc-info-get-help-option cmd-args)
    (bmc-info-display-help))
   ((bmc-info-get-usage-option cmd-args)
    (bmc-info-display-usage))
   ((bmc-info-get-version-option cmd-args)
    (bmc-info-display-version))
   (else 
    (and (fi-ipmi-open cmd-args)
	 (begin (fi-cmd-get-device-id-display)
		(fi-ipmi-close))))))

(define (bmc-info args)
  "fish bmc-info main"
  (let ((cmd-args (bmc-info-argp (append (list "bmc-info") 
					 (list->strlist args)))))
    (if (list? cmd-args)
	(bmc-info-main cmd-args))))

(fi-register-command! 
 (list "bmc-info" 
       (string-append 
	"bmc-info [--no-probing] [--driver-type=IPMIDRIVER]\n"
	"         [--driver-address=DRIVERADDR] [--driver-device=DEVICE]\n"
        "         [--packet-retry-timeout=TIMEOUT] [--packet-retry-max=COUNT]\n"
	"         [--hostname=IPMIHOST] [--username=USERNAME]\n"
	"         [--password=PASSWORD] [--auth-type=AUTHTYPE]\n"
	"         [--priv-level=PRIVILEGE-LEVEL] [--help] [--usage] [--version]\n"
	"\n"
	"   Displays information about BMC.")))

